infer.py

```diff
@@ -18,30 +18,183 @@
 from seq2struct.utils import registry
 from seq2struct.utils import saver as saver_mod
 
+class Inferer:
+    def __init__(self, config):
+        self.config = config
+        if torch.cuda.is_available():
+            self.device = torch.device('cuda')
+        else:
+            self.device = torch.device('cpu')
+            torch.set_num_threads(1)
+
+        # 0. Construct preprocessors
+        self.model_preproc = registry.instantiate(
+            registry.lookup('model', config['model']).Preproc,
+            config['model'])
+        self.model_preproc.load()
+
+    def load_model(self, logdir, step):
+        '''Load a model (identified by the config used for construction) and return it'''
+        # 1. Construct model
+        model = registry.construct('model', self.config['model'], preproc=self.model_preproc, device=self.device)
+        model.to(self.device)
+        model.eval()
+        model.visualize_flag = False
+
+        optimizer = registry.construct('optimizer', self.config['optimizer'], params=model.parameters())
+
+        # 2. Restore its parameters
+        saver = saver_mod.Saver(model, optimizer)
+        last_step = saver.restore(logdir, step=step, map_location=self.device)
+        if not last_step:
+            raise Exception('Attempting to infer on untrained model')
+        return model
+
+    def infer(self, model, output_path, args):
```

Comment:
```suggestion
    def run(self, model, output_path, args):
```

Reply:
Did you want to change Trainer.train() to Trainer.run(), and Preprocessor.preprocess() to Preprocessor.run() also?

Reply:
I think it's fine either way. For `infer`, it seemed undesirable to have both `infer` and `_infer`.

Reply:
Okay. Then for consistency, if you don't mind, I'd leave that as infer() and rename _infer() to be something like _innerInfer (or, a better name if you have a suggestion?).